API: Restrict anonymous user information access

marefr opened the pull request:
**What this PR does / why we need it**:
Existing /api/alert-notifications now requires at least editor access.
Existing /api/alert-notifiers now requires at least editor access.
New /api/alert-notifications/lookup returns less information than
/api/alert-notifications and can be access by any authenticated user.

Existing /api/org/users now requires org admin role.
New /api/org/users/lookup returns less information than
/api/org/users and can be access by users that are org admins,
admin in any folder or admin of any team.
UserPicker component now uses /api/org/users/lookup instead
of /api/org/users.

**Which issue(s) this PR fixes**:
Fixes #17318 

**Special notes for your reviewer**:
/api/org/users/lookup checks if user is org admin or admin of any folder or admin in any team - not sure if this is correct?

marefr committed:
alerting: restrict alert notifications access

Existing /api/alert-notifications now requires at least editor access.
Existing /api/alert-notifiers now requires at least editor access.
New /api/alert-notification-channels returns less information than
/api/alert-notifications and can be access by any authenticated user.

marefr commented on pkg/api/api.go:
```diff
@@ -343,8 +343,9 @@ func (hs *HTTPServer) registerRoutes() {
 
-		apiRoute.Get("/alert-notifications", Wrap(GetAlertNotifications))
-		apiRoute.Get("/alert-notifiers", Wrap(GetAlertNotifiers))
+		apiRoute.Get("/alert-notification-channels", Wrap(GetAlertNotificationChannels))
```
The http api for alert notifications/channels is totally messed up so didn't now what to call this new endpoint. Any better suggestions/ideas?

torkelo replied:
feels a bit messy to have it as a new top level URI resource name. 
* /alert-notifications/read-only
* /alert-notifications/basic-list

marefr replied:
Yeah, I totally agree. For the /org/users I'm thinking of adding a `/org/users/lookup` so maybe `/alert-notifications/lookup` can work?

torkelo replied:
yes!

marefr commented on pkg/api/api.go:
```diff
@@ -343,8 +343,9 @@ func (hs *HTTPServer) registerRoutes() {
-		apiRoute.Get("/alert-notifiers", Wrap(GetAlertNotifiers))
+		apiRoute.Get("/alert-notification-channels", Wrap(GetAlertNotificationChannels))
+		apiRoute.Get("/alert-notifications", reqEditorRole, Wrap(GetAlertNotifications))
+		apiRoute.Get("/alert-notifiers", reqEditorRole, Wrap(GetAlertNotifiers))
```
Noticed this endpoint is only used when editing alerting notifcation channels so required at least editor for that one. Can probable move down the GET /alert-notifications to the /alert-notifications group.

marefr commented on pkg/api/dtos/alerting.go:
```diff
@@ -77,6 +77,24 @@ type AlertNotification struct {
+	}
+}
+
+type AlertNotificationChannel struct {
```
Created a new struct instead of reusing the `AlertNotification` struct. Makes it more explicit and harder to make mistakes.

torkelo replied:
not sure I like having AlertNotification and AlertNotificationChannel dtos that are the same thing without any clear way to know what makes them different. Easy to iintoduce mistakes / leaks in the future.
ideas:
* AlertNotificationChannelFull 
* AlertNotificationChannelStripped

* AlertNotificationChannelForEditor
* AlertNotificationChannelForViewer

marefr replied:
Given https://github.com/grafana/grafana/pull/18422#discussion_r311410954 I would rename this to AlertNotificationLookup.

torkelo replied:
yes!

marefr committed:
alerting: restrict org users access

Existing /api/org/users now requires org admin role.
New /api/org/users/lookup returns less information than
/api/org/users and can be access by users that are org admins,
admin in any folder or admin of any team.
UserPicker component now uses /api/org/users/lookup instead
of /api/org/users.

marefr committed:
alerting: alert notifications access fixes after review

marefr committed:
chore: remove aliased imports

papagian commented on pkg/api/api.go:
```diff
@@ -343,10 +344,10 @@ func (hs *HTTPServer) registerRoutes() {
+		apiRoute.Get("/alert-notifiers", reqEditorRole, Wrap(GetAlertNotifiers))
 
 		apiRoute.Group("/alert-notifications", func(alertNotifications routing.RouteRegister) {
+			apiRoute.Get("/", Wrap(GetAlertNotifications))
```
not sure but this shouldn't be:
`alertNotifications.Get("/", Wrap(GetAlertNotifications))`?

marefr replied:
Good catch, will fix :+1:

papagian reviewed (changes requested):
Great work!
Only [this](https://github.com/grafana/grafana/pull/18422#discussion_r312355780) small fix is required.
Also these docs need to be updated:
https://github.com/grafana/grafana/blob/master/docs/sources/http_api/alerting_notification_channels.md
https://github.com/grafana/grafana/blob/master/docs/sources/http_api/org.md

marefr committed:
use correct route register for alert notifications lookup

marefr committed:
docs: update alert notifications and org  http api

marefr committed:
fix spelling error

papagian commented on docs/sources/http_api/org.md:
```diff
@@ -47,6 +47,9 @@ Content-Type: application/json
 
 `GET /api/org/users`
 
+Retuens all org users within the current organization.
```
spelling error

papagian commented on docs/sources/http_api/org.md:
```diff
@@ -64,11 +67,47 @@ Content-Type: application/json
+
+`GET /api/org/users/lookup`
+
+Retuens all org users within the current organization, but with less detailed information.
```
spelling error

papagian reviewed (changes requested):
There are some misspellings in the docs.

Also, make sure that `api/org/users/lookup` should not be used instead in the following files:
https://github.com/grafana/grafana/blob/master/public/app/features/users/state/actions.ts#L48
https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/gettingstarted/GettingStarted.tsx#L82

and `/api/alert-notifications/lookup` in these:
https://github.com/grafana/grafana/blob/master/public/app/features/alerting/NotificationsEditCtrl.ts#L90
https://github.com/grafana/grafana/blob/master/public/app/features/alerting/NotificationsListCtrl.ts#L15

marefr committed:
fix spelling error

marefr commented:
> There are some misspellings in the docs.
> 
> Also, make sure that `api/org/users/lookup` should not be used instead in the following files:
> https://github.com/grafana/grafana/blob/master/public/app/features/users/state/actions.ts#L48
> https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/gettingstarted/GettingStarted.tsx#L82
> 
> and `/api/alert-notifications/lookup` in these:
> https://github.com/grafana/grafana/blob/master/public/app/features/alerting/NotificationsEditCtrl.ts#L90
> https://github.com/grafana/grafana/blob/master/public/app/features/alerting/NotificationsListCtrl.ts#L15

**Won't fix - they're correct as far as I'm concern:**
https://github.com/grafana/grafana/blob/master/public/app/features/users/state/actions.ts#L48
https://github.com/grafana/grafana/blob/master/public/app/features/alerting/NotificationsEditCtrl.ts#L90 https://github.com/grafana/grafana/blob/master/public/app/features/alerting/NotificationsListCtrl.ts#L15

**Fixed:**
https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/gettingstarted/GettingStarted.tsx#L82

marefr committed:
fix getting started panel

papagian commented:
Great!
